reactor-core/src/main/java/reactor/core/publisher/MonoTakeLastOne.java

```diff
@@ -56,74 +55,133 @@ public Object scanUnsafe(Attr key) {
-		Subscription s;
+
+		T value;
+
+		boolean done;
 
 		TakeLastOneSubscriber(CoreSubscriber<? super T> actual,
 				@Nullable T defaultValue,
 				boolean mustEmit) {
 			super(actual);
-			this.defaultValue = defaultValue;
+			this.value = defaultValue;
 			this.mustEmit = mustEmit;
 		}
 
-		@Override
-		public void onSubscribe(Subscription s) {
-			if (Operators.validate(this.s, s)) {
-				this.s = s;
-
-				actual.onSubscribe(this);
-
-				s.request(Long.MAX_VALUE);
-			}
-
-		}
-
 		@Override
 		@Nullable
 		public Object scanUnsafe(Attr key) {
-			if (key == Attr.PARENT) return s;
-			if (key == Attr.RUN_STYLE) return Attr.RunStyle.SYNC;
+			if (key == Attr.TERMINATED) return done && value == null;
+			if (key == Attr.CANCELLED) return value == CANCELLED;
 
 			return super.scanUnsafe(key);
 		}
 
 		@Override
 		public void onNext(T t) {
 			T old = this.value;
-			setValue(t);
+			if (old == CANCELLED) {
+				// cancelled
+				Operators.onDiscard(t, actual.currentContext());
+				return;
+			}
+
+			synchronized (this) {
+				old = this.value;
+				if (old != CANCELLED) {
+					this.value = t;
+				}
+			}
+
+			if (old == CANCELLED) {
+				// cancelled
+				Operators.onDiscard(t, actual.currentContext());
+				return;
+			}
+
 			Operators.onDiscard(old, actual.currentContext()); //FIXME cache context
 		}
 
+		@Override
+		public void onError(Throwable t) {
+			if (this.done) {
+				Operators.onErrorDropped(t, actual.currentContext());
+				return;
+			}
+			this.done = true;
+
+
+			final T v;
+			synchronized (this) {
+				v = this.value;
+				this.value = null;
+			}
+
+			if (v == CANCELLED) {
+				Operators.onErrorDropped(t, actual.currentContext());
+				return;
+			}
+
+			this.actual.onError(t);
+		}
+
 		@Override
 		public void onComplete() {
-			T v = this.value;
+			if (this.done) {
+				return;
+			}
+			this.done = true;
+
+			final T v = this.value;
+
+			if (v == CANCELLED) {
+				return;
+			}
+
 			if (v == null) {
 				if (mustEmit) {
-					if(defaultValue != null){
-						complete(defaultValue);
-					}
-					else {
-						actual.onError(Operators.onOperatorError(new NoSuchElementException(
-								"Flux#last() didn't observe any " + "onNext signal"),
-								actual.currentContext()));
-					}
+					actual.onError(Operators.onOperatorError(new NoSuchElementException(
+							"Flux#last() didn't observe any " + "onNext signal"),
+							actual.currentContext()));
 				}
 				else {
 					actual.onComplete();
 				}
 				return;
 			}
-			complete(v);
+
+			completeWhenEmpty();
 		}
 
 		@Override
 		public void cancel() {
-			super.cancel();
 			s.cancel();
+
+			final T v;
+			synchronized (this) {
+				v = this.value;
+				//noinspection unchecked
+				this.value = (T) CANCELLED;
+			}
+
+			if (v != null) {
+				Operators.onDiscard(v, actual.currentContext());
+			}
+		}
+
+		@Override
+		T resolveValue() {
+			final T v;
+			synchronized (this) {
+				v = this.value;
+				this.value = null;
+			}
+			return v;
```

Comment:
yeah... bug

Reply:
☑️

Reply:
`MonoReduce` also has a `CANCELLED` constant but isn't susceptible to this because it doesn't extend `BaseFluxToMonoOperator`. other child classes don't have the issue. resolving this particular conversation 👍